gpu/src/cuda/shaders/fp_u256.cuh

```diff
@@ -0,0 +1,210 @@
+// https://github.com/andrewmilson/ministark/blob/main/gpu-poly/src/metal/felt_u256.h.metal
+
+#ifndef felt_u256_h
+#define felt_u256_h
+
+#include "u256.cuh"
+
+template <
+    /* =N **/ unsigned long N_0, unsigned long N_1, unsigned long N_2,
+    unsigned long N_3,
+    /* =R_SQUARED **/ unsigned long R_SQUARED_0, unsigned long R_SQUARED_1,
+    unsigned long R_SQUARED_2, unsigned long R_SQUARED_3,
+    /* =N_PRIME **/ unsigned long N_PRIME_0, unsigned long N_PRIME_1,
+    unsigned long N_PRIME_2, unsigned long N_PRIME_3>
+class Fp256 {
+public:
+  Fp256() = default;
+  __device__ constexpr Fp256(unsigned long v) : inner(v) {}
+  __device__ constexpr Fp256(u256 v) : inner(v) {}
+
+  __device__ constexpr explicit operator u256() const { return inner; }
+
+  __device__ constexpr Fp256 operator+(const Fp256 rhs) const {
+    return Fp256(add(inner, rhs.inner));
+  }
+
+  __device__ constexpr Fp256 operator-(const Fp256 rhs) const {
+    return Fp256(sub(inner, rhs.inner));
+  }
+
+  __device__ Fp256 operator*(const Fp256 rhs) const {
+    return Fp256(mul(inner, rhs.inner));
+  }
+
+  // TODO: make method for all fields
+  __device__ Fp256 pow(unsigned exp) {
+    // TODO find a way to generate on compile time
+    Fp256 const ONE = mul(u256(1), R_SQUARED);
+    Fp256 res = ONE;
+
+    while (exp > 0) {
+      if (exp & 1) {
+        res = res * *this;
+      }
+      exp >>= 1;
+      *this = *this * *this;
+    }
+
+    return res;
+  }
+
+  __device__ Fp256 inverse() {
+    // used addchain
+    // https://github.com/mmcloughlin/addchain
```

Comment:
Neat trick! I'll probably steal it for cairo 👀

Reply:
Be my guest!